### PR TITLE
fix crash when the expression exceed the depth

### LINKED ISSUE #3393: Limit the depth of the expression, report an error when it exceeds but not crash

### PR DESCRIPTION
**Pull request by heroicNeZha**
#### What type of PR is this?
- [x] bug
- [ ] feature
- [ ] enhancement

#### What does this PR do?
closes #3393

#### Which issue(s)/PR(s) this PR relates to?
```
# when the expression exceed the depth
[ERROR (-1004)]: SyntaxError: The above expression is not a valid expression, because its depth exceeds the maximum depth! near `1+1+1+1+1+1+1+1+1+1+1+1+1+1+1+1+1+1+1+1+1+1+1+1+1+1+1+1+1+1+1+1+1+1+1+1+1+1+1+1+'
```
calcate the depth in parser through BFS
#### Checklist:
- [ ] Documentation affected (Please add the label if documentation needs to be modified.)
- [ ] Incompatibility (If it breaks the compatibility, please describe it and add the corresponding label.）
- [ ] If it's needed to cherry-pick (If cherry-pick to some branches is required, please label the destination version(s).)
- [ ] Performance impacted: Consumes more CPU/Memory

#### Release notes:

Please confirm whether to be reflected in release notes and how to describe:
>                                                                 `

### CODE REVIEW

File: src/parser/parser.yy [hunk 1]
```diff
@@ -555,6 +555,15 @@ unreserved_keyword
     ;
 
 expression
+    : expression_internal {
+        $$ = $1;
+        if(!graph::ExpressionUtils::checkExprDepth($$)){
+            throw nebula::GraphParser::syntax_error(@1, "The above expression's depth exceeds the maximum depth!");
```
Comment (yixinglu) [inline src/parser/parser.yy:561]: need to delete this $$ before throw exception

Comment (yixinglu) [reply src/parser/parser.yy:561]: The above expression's depth exceeds the maximum depth: 512!

ExpressionUtils::kMaxDepth

Comment (heroicNeZha) [reply src/parser/parser.yy]: ok

**Review by yixinglu — APPROVED** [review]

File: src/graph/util/ExpressionUtils.cpp [hunk 2]
```diff
@@ -984,5 +986,180 @@ bool ExpressionUtils::isGeoIndexAcceleratedPredicate(const Expression *expr) {
...
+        case Expression::Kind::kTSFuzzy:
+        case Expression::Kind::kAggregate:
+        case Expression::Kind::kSubscriptRange:
+        case Expression::Kind::kVersionedVar:
+        default:
+          break;
+      }
+      size -= 1;
+    }
+    depth += 1;
```
Comment (czpmango) [inline src/graph/util/ExpressionUtils.cpp:1159]: if (depth > kMaxDepth) return false;

**Review by jievince — DISMISSED** [review]
LGTM

File: src/graph/util/ExpressionUtils.cpp (hunk 2 above)
Comment (heroicNeZha) [reply src/graph/util/ExpressionUtils.cpp:1159]: fixed

**Review by heroicNeZha — COMMENTED** [review]
# without depth check
============================================================================
/home/nebula/nebula/src/parser/test/ParserBenchmark.cpprelative  time/iter  iters/s
============================================================================
SimpleQuery(1_thread)                                       16.69us   59.93K
SimpleQuery(2_thread)                             99.27%    16.81us   59.49K
SimpleQuery(4_thread)                             99.38%    16.79us   59.56K
SimpleQuery(8_thread)                             99.17%    16.83us   59.43K
SimpleQuery(16_thread)                            99.00%    16.86us   59.33K
SimpleQuery(32_thread)                            95.46%    17.48us   57.21K
SimpleQuery(48_thread)                            86.43%    19.31us   51.80K
----------------------------------------------------------------------------
ComplexQuery(1_thread)                                      75.65us   13.22K
ComplexQuery(2_thread)                            99.68%    75.90us   13.18K
ComplexQuery(4_thread)                            99.51%    76.03us   13.15K
ComplexQuery(8_thread)                            99.51%    76.02us   13.15K
ComplexQuery(16_thread)                           98.88%    76.51us   13.07K
ComplexQuery(32_thread)                           95.30%    79.38us   12.60K
ComplexQuery(48_thread)                           85.81%    88.17us   11.34K
============================================================================
# add depth check
============================================================================
/home/nebula/nebula/src/parser/test/ParserBenchmark.cpprelative  time/iter  iters/s
============================================================================
SimpleQuery(1_thread)                                       16.49us   60.65K
SimpleQuery(2_thread)                             99.38%    16.59us   60.27K
SimpleQuery(4_thread)                             99.22%    16.62us   60.17K
SimpleQuery(8_thread)                             99.20%    16.62us   60.16K
SimpleQuery(16_thread)                            98.25%    16.78us   59.58K
SimpleQuery(32_thread)                            95.33%    17.30us   57.81K
SimpleQuery(48_thread)                            85.46%    19.29us   51.83K
----------------------------------------------------------------------------
ComplexQuery(1_thread)                                      79.23us   12.62K
ComplexQuery(2_thread)                            99.81%    79.38us   12.60K
ComplexQuery(4_thread)                            99.88%    79.32us   12.61K
ComplexQuery(8_thread)                            99.85%    79.35us   12.60K
ComplexQuery(16_thread)                           99.28%    79.80us   12.53K
ComplexQuery(32_thread)                           96.18%    82.37us   12.14K
ComplexQuery(48_thread)                           84.38%    93.90us   10.65K
============================================================================

**Review by yixinglu — DISMISSED** [review]

**Review by jievince — DISMISSED** [review]

**Review by jievince — APPROVED** [review]

**Review by CPWstatic — APPROVED** [review]

**Review by yixinglu — APPROVED** [review]

**Pull request merged**